lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3352,15 +3352,38 @@ namespace {
         }
       }
 
+      bool retunrsRetainedUnretainedDiagnosticsNeeded = false;
+      if (!isa<clang::CXXDeductionGuideDecl>(decl)) {
+        if (const auto *methodDecl = dyn_cast<clang::CXXMethodDecl>(decl)) {
+          if (isNormalCxxMemberMethod(methodDecl)) {
+            // normal c++ member method
+            retunrsRetainedUnretainedDiagnosticsNeeded = true;
+          }
+        } else {
+          // global or top-level function
+          retunrsRetainedUnretainedDiagnosticsNeeded = true;
+        }
+      }
+
       HeaderLoc loc(decl->getLocation());
       if (isForeignReferenceTypeWithoutImmortalAttrs(decl->getReturnType())) {
         if (returnsRetainedAttrIsPresent && returnsUnretainedAttrIsPresent) {
           Impl.diagnose(loc, diag::both_returns_retained_returns_unretained,
                         decl);
         } else if (!returnsRetainedAttrIsPresent &&
                    !returnsUnretainedAttrIsPresent) {
-          Impl.diagnose(loc, diag::no_returns_retained_returns_unretained,
-                        decl);
+          if (retunrsRetainedUnretainedDiagnosticsNeeded) {
```

Comment:
imo the `retunrsRetainedUnretainedDiagnosticsNeeded` check should only performed when `!returnsRetainedAttrIsPresent && !returnsUnretainedAttrIsPresent`, so it should be done right here. The details (including all the compat/versioning stuff) can be encapsulated within a helper function, as Gabor suggests, e.g.,

```cpp
if (!returnsRetainedAttrIsPresent && !returnsUnretainedAttrIsPresent) {
  diagnoseUnannotatedFunctionReturningFRT(/* etc */);
}
```